packages/flutter/lib/src/widgets/heroes.dart

```diff
@@ -849,40 +849,48 @@ class HeroController extends NavigatorObserver {
     HeroFlightDirection flightType,
     bool isUserGestureTransition,
   ) {
-    if (toRoute != fromRoute && toRoute is PageRoute<dynamic> && fromRoute is PageRoute<dynamic>) {
-      final PageRoute<dynamic> from = fromRoute;
-      final PageRoute<dynamic> to = toRoute;
+    if (toRoute == fromRoute) {
```

Comment:
this just refactor the if condition to return early instead of nested condition.

the real functional change is in line 889 where it check for whether the route is still active before actually do the fly.